kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergWriterCompatMetadataValidatorAndUpdater.java

```diff
@@ -0,0 +1,257 @@
+        throw DeltaErrors.icebergCompatIncompatibleTableFeatures(
+            compatFeatureName, Collections.singleton(INVARIANTS_W_FEATURE));
+      }
+    };
+  }
+
+  /**
+   * Checks that the table feature `changeDataFeed` is not active in the table, meaning the table
+   * property `delta.enableChangeDataFeed` is not enabled.
+   */
+  protected static IcebergCompatCheck createChangeDataFeedInactiveCheck(String compatFeatureName) {
+    // Note - since Kernel currently does not support the table feature `changeDataFeed` we will
+    // not hit this check for E2E writes since we will fail early due to unsupported write
+    // If Kernel starts supporting the feature `changeDataFeed` this check will become applicable
+    return (inputContext) -> {
+      if (TableConfig.CHANGE_DATA_FEED_ENABLED.fromMetadata(inputContext.newMetadata)) {
+        throw DeltaErrors.icebergCompatIncompatibleTableFeatures(
+            compatFeatureName, Collections.singleton(CHANGE_DATA_FEED_W_FEATURE));
+      }
+    };
+  }
+
+  /**
+   * Checks that the table feature `checkConstraints` is not active in the table, meaning the table
+   * has no check constraints stored in its metadata configuration.
+   */
+  protected static IcebergCompatCheck createCheckConstraintsInactiveCheck(
+      String compatFeatureName) {
+    // Note - since Kernel currently does not support the table feature `checkConstraints` we will
+    // not hit this check for E2E writes since we will fail early due to unsupported write
+    // If Kernel starts supporting the feature `checkConstraints` this check will become
+    // applicable
+    return (inputContext) -> {
+      if (TableFeatures.hasCheckConstraints(inputContext.newMetadata)) {
+        throw DeltaErrors.icebergCompatIncompatibleTableFeatures(
+            compatFeatureName, Collections.singleton(CONSTRAINTS_W_FEATURE));
+      }
+    };
+  }
+
+  /**
+   * Checks that the table feature `identityColumns` is not active in the table, meaning no identity
+   * columns exist in the table schema.
+   */
+  protected static IcebergCompatCheck createIdentityColumnsInactiveCheck(String compatFeatureName) {
+    // Note - since Kernel currently does not support the table feature `identityColumns` we will
+    // not hit this check for E2E writes since we will fail early due to unsupported write
+    // If Kernel starts supporting the feature `identityColumns` this check will become applicable
+    return (inputContext) -> {
+      if (TableFeatures.hasIdentityColumns(inputContext.newMetadata)) {
+        throw DeltaErrors.icebergCompatIncompatibleTableFeatures(
+            compatFeatureName, Collections.singleton(IDENTITY_COLUMNS_W_FEATURE));
+      }
+    };
+  }
+
+  /**
+   * Checks that the table feature `generatedColumns` is not active in the table, meaning no
+   * generated columns exist in the table schema.
+   */
+  protected static IcebergCompatCheck createGeneratedColumnsInactiveCheck(
```

Comment:
why create these checks instead declare them once and reuse those objects? Just like how it is currently in the IcebergWriterCompatV2MetadataValidatorAndUpdater